### PR TITLE
Fix security user creation

### PR DESCRIPTION
**Pull request by Ninjeneer**

### CODE REVIEW

File: src/assets/server/rest/v1/schemas/common/common.json [hunk 1]
```diff
@@ -112,7 +112,8 @@
     },
     "password": {
       "type": "string",
-      "sanitize": "mongo"
+      "sanitize": "mongo",
+      "pattern": "(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!#@:;,<>\/''\\$%\\^&\\*\\.\\?\\-_\\+\\=\\(\\)])(?=.{8,})"
```
Comment (jerome-benoit) [inline src/assets/server/rest/v1/schemas/common/common.json:116]: Not sure the pattern enforce a min length, have you checked?

**Review by jerome-benoit — CHANGES_REQUESTED** [review]
I probably more need explanation than changes ...

Thanks for taking care of it.

File: src/server/rest/v1/service/UtilsService.ts [hunk 2]
```diff
@@ -1324,7 +1324,7 @@ export default class UtilsService {
         actionOnUser: filteredRequest.id
       });
     }
-    if (filteredRequest.password && !Utils.isPasswordValid(filteredRequest.password)) {
+    if ((req.method === 'POST' && (!filteredRequest.password || !Utils.isPasswordValid(filteredRequest.password))) || filteredRequest.password && !Utils.isPasswordValid(filteredRequest.password)) {
```
[the remark this answers is not in the document]
Comment (Ninjeneer) [reply src/server/rest/v1/service/UtilsService.ts:1327]: I made this to handle the difference between user creation and update

User creation (POST) requires a non-null valid password
User update (PUT) does not require password, but can have one

File: src/assets/server/rest/v1/schemas/common/common.json (hunk 1 above)
Comment (Ninjeneer) [reply src/assets/server/rest/v1/schemas/common/common.json:116]: Yes it does

File: src/server/rest/v1/service/UtilsService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply src/server/rest/v1/service/UtilsService.ts:1327]: Fair enough, can you just add a comment before the test stating it? 

Thanks.

File: src/assets/server/rest/v1/schemas/common/common.json (hunk 1 above)
Comment (jerome-benoit) [reply src/assets/server/rest/v1/schemas/common/common.json:116]: And the validator does not does not barf in the PUT case if the password is empty?

**Review by jerome-benoit — CHANGES_REQUESTED** [review]

File: src/server/rest/v1/service/UtilsService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ninjeneer) [reply src/server/rest/v1/service/UtilsService.ts:1327]: Ok

File: src/assets/server/rest/v1/schemas/common/common.json (hunk 1 above)
Comment (Ninjeneer) [reply src/assets/server/rest/v1/schemas/common/common.json:116]: There is not yet a validator for the user creation/update, it is still handled by the UserSecurity.
But it won't, as we could set the 'password' field as not required in the update scheme

**Review by jerome-benoit — APPROVED** [review]
Good to go. I really hope that all passwords in DB match the requirements ;)

File: src/server/rest/v1/service/UtilsService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LucasBrazi06) [reply src/server/rest/v1/service/UtilsService.ts:1328]: Non readable IF statement, split it in 2: One for Create and one for Update.
Move the check of the filteredRequest.password in the Utils method.
Then I will not frown my eyebrows ;-)

**Review by jerome-benoit — APPROVED** [review]

**Comment by jerome-benoit** [comment]
@LucasBrazi06: I took the liberty to merge that PR during your vacation, the security issue is critical. I will probably also deploy it in production if I have to right to do it ...

**Pull request merged**